crates/bpe/README.md

```diff
@@ -0,0 +1,168 @@
+
+Any substring `e_i..e_j` of a valid encoding sequence `e_0..e_n` is again a valid encoding sequence (with `0 <= i <= j <= n`).
+
+This is obviously true, since the BPE algorithm didn't merge any characters across the token boundaries of the valid encoding sequence `e_0..e_n`.
+This means that the BPE algorithm would execute exactly the same merge operations for the substring `e_i..e_j` as for the full sequence even if it only sees that substring.
+
+### Corollary IIa
+
+Knowing the last (or first) token of a valid encoding sequence uniquely determines the full encoding sequence.
+
+This is a direct consequence of corollary II. Given a non-empty valid encoding sequence `e_0..e_n` for some input text `T`. Removing the last token `e_n` leads to a shorter valid encoding sequence according to corollary II which corresponds to the input text `T'` after removing the last bytes represented by the last token.
+The same argument can now be applied to the shorter sequence `e_0..e_{n-1}` and the text `T'` until the full sequence is reconstructed.
+
+### Corollary III
+
+Given two valid encoding sequences `e_0..e_i` and `e_i..e_n`, then `e_0..e_i..e_n` is also a valid encoding sequence.
+Note: that the end/start token has to be identical between the two sequences!
+
+The correctness of this statement follows with a similar argument as used in Corollary II.
+Given the merge operations performed by BPE for both valid encoding sequences. The merge operations which lead to the shared token `e_i` must be identical to produce the same token. And those are the only redundant merge operations. Combining the two sets of merge operations will lead to the combined token sequence.
+If BPE wants to make a different merge decision when it sees the full input, then this merge decision must involve either the token boundary to the left or right of `e_i`. But that means that it had to make a different merge decision for at least one of the substrings `e_0..e_i` or `e_i..e_n` which cover those token boundaries. So, by contradiction, the corollary must be true.
+
+### Corollary IIIa
+
+Given a valid encoding sequence `e_0..e_i` and a valid encoding tuple `e_i e_j`, then `e_0..e_i e_j` is also a valid encoding sequence.
+
+
+## Novel Algorithm
+
+At a first glance, it seems impossible to achieve `O(n)` complexity while preserving the encoding output of the original BPE algorithm, since the original BPE algorithm needs to first scan the full input before it can make any encoding decision.
+For instance, the sequence `abab` would be encoded as `ab ab` when the dictionary contains the tokens `a b ab ba bc abc babc ababc` ordered by frequency. But appending a single character `ababc` would result in a pretty different tokenization: `ab a bc`. So without looking ahead it seems impossible to properly tokenize the text.
+
+The solution is to track ALL encodings for all text prefixes. For our example `ababc` we would get:
+- `a` ------> `a`
+- `ab` -----> `ab`
+- `aba` ----> `ab a`
```

Comment:
**Question:** Why is the resulting encoding `ab a` rather than `a ba`? Is this the consequence of

> Then, one scans over the preliminary encoding and determines the smallest token id by which any pair of neighboring token ids could be replaced.
The left most of them is replaced with that token id.

I guess why is the single `a` token not considered the smallest token id in this case, given its position in the dictionary implies highest frequency?

Reply:
If you want to encode `aba`, then you start with the byte tokens `a` `b` `a`.
Next, you look for potential merges. in this example you can choose between `ab` and `ba`.
Since `ab` is the smaller one of the two you replace that one first.
I.e. you get now `ab` `a`.
But at this point not possible merge operation exists anymore (`ba` is no longer possible)

Reply:
> Next, you look for potential merges. in this example you can choose between ab and ba.
Since ab is the smaller one of the two you replace that one first.

Thanks for clarifying, I understand now. That merge operation is producing the smallest token id for a sequence of individual byte tokens or already paired tokens starting with the left most element working towards the end of the sequence.